packages/access-api/src/service/upload-api-proxy.js

```diff
@@ -0,0 +1,266 @@
+  static createConnection(options) {
+    return Client.connect({
+      id: dagUcanDid.parse(options.audience),
+      encoder: CAR,
+      decoder: CBOR,
+      channel: HTTP.open({
+        fetch: options.fetch,
+        url: options.url,
+      }),
+    })
+  }
+
+  /**
+   * @param {object} options
+   * @param {Record<dagUcan.DID, URL>} options.audienceToUrl
+   * @param {UcantoHttpConnectionOptions} [options.defaultConnection]
+   * @param {typeof globalThis.fetch} options.fetch
+   */
+  constructor(options) {
+    this.#fetch = options.fetch
+    this.#audienceToUrl = options.audienceToUrl
+    this.#defaultConnection = options.defaultConnection
+      ? AudienceConnections.createConnection(options.defaultConnection)
+      : undefined
+  }
+
+  /**
+   * @param {dagUcan.DID} audience
+   */
+  get(audience) {
+    if (audience in this.#audienceToUrl) {
+      return AudienceConnections.createConnection({
+        audience,
+        fetch: this.#fetch,
+        url: this.#audienceToUrl[audience],
+      })
+    }
+    return this.#defaultConnection
+  }
+}
+
+export class UploadApiProxyService {
+  /** @type {StoreService} */
+  store
+  /** @type {UploadService} */
+  upload
+
+  /**
+   * @type {Record<string, {
+   * audience: dagUcan.DID,
+   * url: URL,
+   * }>}
+   */
+  static #environments = {
+    production: {
+      audience: 'did:web:web3.storage',
+      url: new URL('https://up.web3.storage'),
+    },
+    staging: {
+      audience: 'did:web:staging.web3.storage',
+      url: new URL('https://staging.up.web3.storage'),
+    },
+  }
+
+  /**
+   * @param {object} options
+   * @param {iucanto.Signer} [options.signer]
+   * @param {typeof globalThis.fetch} options.fetch
+   */
+  static create(options) {
+    // eslint-disable-next-line unicorn/no-array-reduce
+    const audienceToUrl = Object.values(this.#environments).reduce(
+      (acc, { audience, url }) => {
+        acc[audience] = url
+        return acc
+      },
+      /** @type {Record<dagUcan.DID, URL>} */ ({})
+    )
+    const connections = new AudienceConnections({
+      audienceToUrl,
+      defaultConnection: {
+        fetch: options.fetch,
+        ...this.#environments.production,
+      },
+      fetch: options.fetch,
+    })
+    const proxyOptions = {
+      signer: options.signer,
+      connections,
+    }
+    return new this(
+      createProxyStoreService(proxyOptions),
+      createProxyUploadService(proxyOptions)
+    )
+  }
+
+  /**
+   * @protected
+   * @param {StoreService} store
+   * @param {UploadService} upload
+   */
+  constructor(store, upload) {
+    this.store = store
+    this.upload = upload
+  }
+}
+
+/**
+ * Return whether the provided stack trace string appears to be generated
+ * by a deployed upload-api.
+ * Heuristics:
+ * * stack trace files paths will start with `file:///var/task/upload-api` because of how the lambda environment is working
+ *
+ * @param {string} stack
+ */
+export function isUploadApiStack(stack) {
```

Comment:
Nit: Having this here makes me uncomfortable, seems to me that it belongs in tests so it's not used for other things.

Reply:
https://github.com/web3-storage/w3protocol/pull/334/commits/49144adb8f52904af020a3b3db096b42d69e3f65